Blake

jeremydaily opened the pull request.

Fusickb committed:
Create README.md

Created tcpCANClient.py to collect data from a CAN interface and send it to a TCP Server. tcpCANServer.py listens for the client and formats the data of the CAN frames to the console similar to the formatting of readRawCANSocket.py.

Fusickb committed:
Create tcpCANServer.py

Upload of tcpCANServer.

Fusickb committed:
Create tcpCANClient.py

Upload of tcpCANClient.py.

Fusickb committed:
Update tcpCANClient.py

Change max CAN frames per TCP packet from 93 to 89 and max payload size from 1489 to 1425 to account for header fields of IP and TCP. Allows for up to 75 bytes for headers without fragmentation of messages.

Fusickb committed:
Update tcpCANServer.py

Previous fix from 1489 bytes to 1425 bytes max size of payload.

Fusickb committed:
Update tcpCANServer.py

Removed dependencies on socketCAN so server need not be hosted on a Linux OS. Server now automatically sets the IP address to the machine's IP address.

Fusickb committed:
Update tcpCANClient.py

Client still defaults to loopback address, but comment added to assist others when setting up.

Fusickb committed:
Update tcpCANServer.py

Added logging to text file on server side as formated line for each CAN frame.

Fusickb committed:
Update tcpCANClient.py

Adjusted timeout to 5 seconds for testing.

Fusickb committed:
Update README.md

Updated documentation.

jeremydaily commented:
Make the CANLog file on the server a unique ID with a changeable directory.

Fusickb committed:
Update readRawCANSocket.py

Added socket options to enable socket to receive error frames.

Fusickb committed:
Update tcpCANClient.py

Added socket options to enable socket to receive error frames.

Fusickb committed:
Update readRawCANSocket.py

No real change. File manipulation.

Fusickb committed:
Update tcpCANClient.py

Debugged. RTR, Error, Standard CAN, and Extended CAN frames are now able to be transmitted.

Fusickb committed:
Update tcpCANServer.py

Debugged.

Fusickb committed:
Update readRawCANSocket.py

Added comment

Fusickb committed:
Update tcpCANServer.py

Each CAN log now has a randomly generated universally unique ID generated by the uuid python library. Directory name is now specified so logging directory can be chosen.

Fusickb committed:
Update README.md

Updated documentation

Fusickb committed:
Update README.md

Updated documentation.

Fusickb commented:
Suggested changes made.

Fusickb committed:
Created two files.

Renamed old client/server code to tcpCANrxClient.py and tcpCANrxServer.py for the simple transfer and logging of CAN data to another machine.  tcpCANClient.py and tcpCANServer.py will be repurposed to a more sophisticated control of receiving and transmitting CAN messages.

Fusickb committed:
Started command server/client basic structure.

Created initial structure for command line options to send commands with client and accept requests with server. Intermediate step to accepting commands on the server.

Fusickb committed:
Added connectivity between server and client.

Server now accepts commands as documented in the README. The client accepts command arguments to make this process automatic. The logic for determining which command is performed is complete, but the commands don't yet perform the task.

Fusickb committed:
Update README.md

Documentation fix.

Fusickb committed:
Update tcpCANServer.py

Created logic to start/stop processes based off rxon/rxoff/txon/txoff commands. Up to two additional processes are created for each CAN interface (rx and tx) to host a TCP server on another port (rx) or connect to a TCP server at a specific port (tx).

Fusickb committed:
Implemented rxon/rxoff commands for can0/can1/any interface options.

Created a new rxServers.py file to handle receiving CAN messages over TCP. Implemented functionality where the commands rxon/rxoff actually turn on and off the TCP communications on the node with the tcpCANServer.py script. The functionality is tested and works on can0/can1/any where any is a shortcut for doing all interfaces. Packet drop test at 250k baud NOT performed yet.

Fusickb committed:
Fixed filename bug.

Changed name of file from .txt_interfaceName to interfaceName.txt

Fusickb committed:
Implemented txon/txoff commands for can0/can1/any

Txon/txoff commands now turn on and off servers on different ports on the machine where tcpCANServer.py is found. These commands can now be used with can0/can1/any. An additional cangen command to transmit TCP packets containing CAN frames was also implemented for verification purposes.

jeremydaily commented:
Fix spelling errors in readme file: (e.g. guarunteed should be guaranteed)
Please use `backticks` to highlight commands.
Add an overview section at the top of the readme. It is unclear on which files are used for what purpose. Do we need to run all 4 files? Which scripts run on the different machines?

jeremydaily commented on SocketCAN/rxServers.py:
```diff
@@ -0,0 +1,114 @@
+
+# To match this data structure, the following struct format can be used:
+can_frame_format = "<LL8s"
+# Unsigned Long Integer (little endian), unsigned char (byte), three pad bytes, eight chars (bytes)
```
This comment does not match the frame format.

Fusickb replied:
No, it does not. The second L contains DLC and Micros, which is done on lines 76/77. This was done because there were no format characters from https://docs.python.org/2/library/struct.html with length of 3. Do you suggest this be changed to "<LB3s8s" and the lines 76/77 logic removed?

jeremydaily replied:
You need to update the comments to match the code.

jeremydaily commented on SocketCAN/rxServers.py:
```diff
@@ -0,0 +1,114 @@
+print("socket.CAN_ERR_FLAG = 0x{:08X}".format(socket.CAN_SFF_MASK))"""
+
+CAN_EFF_MASK = 0x1FFFFFFF #assigned to remove socketCAN dependencies, values obtained by above print statements using socketCAN
+CAN_EFF_FLAG = -0x80000000
```
Why is this negative? We should be using unsigned integers.

Fusickb replied:
I'll fix this. It was a straight copy/paste from what was printed when using the CAN_EFF_FLAG from socketCAN. Thanks for the catch.

jeremydaily commented on SocketCAN/rxServers.py:
```diff
@@ -0,0 +1,114 @@
+				try:
+					can_id, can_dlc_and_microTimer, can_data = struct.unpack(can_frame_format, can_packet)
+
+					can_dlc = (can_dlc_and_microTimer & 0x000000FF)  #first byte is dlc b/c little endian
```
"first byte" should be "last byte" in dlc comment.

jeremydaily commented on SocketCAN/tcpCANClient.py:
This does not enable user input like the actual can-utils `cangen`. Let's open an issue and address this in a future pull request.

jeremydaily commented on SocketCAN/tcpCANClient.py:
Add a success message or confirmation.

jeremydaily commented on SocketCAN/tcpCANrxClient.py:
```diff
@@ -0,0 +1,159 @@
+tcpSocket.send(ethData)
+tcpSocket.close()#!/usr/bin/python3
+
+import time
```
Everything below this is duplicated. Please refactor and remove these contents.

Fusickb committed:
Updated documentation and made suggested changes.

Functionally nothing has changed. Some debugging, success messages, changes for correctness.

Fusickb committed:
Updated CAN frame comments. Added command parsing on tcpCANServer.py.

Updated the comments referring to structure of CAN frame. Also added logic for tcpCANServer.py to handle down, up, reset, and setbitrate logic. Functionality not yet implemented.

Fusickb committed:
Implemented functionality of down, up, and reset tcpCANServer.py commands.

Down now turns off specified CAN interface, up now turns on specified CAN interface, and reset now turns off and on specified CAN interface. Any performs those actions on all CAN interfaces within intfOrder. Authentication on the server may be necessary for the first command.

Fusickb committed:
Implemented setbitrate command on tcpCANServer.py.

setbitate command now changes the bitrate of the given CAN interface. The interface any sets the bitrate of all CAN interfaces to the given value. Also added logic for busload, rxmsgs, and txmsgs. They do not have functionality yet.

Fusickb committed:
Implemented busload command.

The busload command now prints to the client the value of the estimated percentage busload gotten from using canbusload on the server side. Updated documentation and added format of percentage busload communication.

Fusickb committed:
Fixed formatting issue of ReadME

Fusickb committed:
Fixed formatting issue on ReadME.

Fusickb committed:
Implemented functionality of rxmsgs and txmsgs tcpCANServer.py commands.

The rxmsgs and txmsgs now display the number of CAN messages received and transmitted over TCP. The results print on the server side and do not transfer the data to the client since it is mostly a feature for packet drop tests and can be observed on either end.
